A usage doc for fly machine run (#1170)
superfly/docs: machines/run.html.md

---

> title: Run a new Machine

Author: I take the point but I'm not sure. That seems like a lot of words. I'm leaving it for now and we can change it later.

The default behavior of `fly machine run` is to create a new Fly App for the new Machine to belong to, unless it's given the name of an existing app in one of two ways:

1. Like many other flyctl commands, `fly machine run` looks for a `fly.toml` app config file in the working directory, and if it finds one with an app name inside, it adds the new Machine to that app. It disregards the rest of the configuration in the file.
2. If you pass it an app name with `--app <app-name>`, flyctl prefers that name over any name it gets from a `fly.toml`.

If the app name doesn't belong to an existing app in one of your orgs, flyctl asks if you want to create it. 

Even if you're not using `fly deploy` to configure and run any of your app's Machines, it may be worth creating a `fly.toml` file with just an app name in it, to save using the `--app` option repeatedly:

```toml
# a fly.toml just to provide an app name to commands 
# run from the same directory

app = <app-name>
```
 
Use `--org <org-name>` to specify which organization a newly created app should belong to. The `--org` flag is ignored when creating the new Machine in an existing app.

## Get or build the Docker image

> All Fly Machines are made from Docker images. When you `fly launch` an app, a Fly Launch scanner may generate one for you based on your source code, or it may be a prebuilt image you point to, or it may be built from your hand-crafted Dockerfile and source code.
> 
> We're working more directly with Machines here, so your choices with `fly machine run` are "point to an image" or "point to a Dockerfile."

Reviewer: I found the introduction of "fly launch"ing an app a bit confusing. You can take or leave whatever from my suggestion below, which attempts to flip around the two scenarios so `fly m run` is first.
```suggestion
All Fly Machines are made from Docker images. When you work directly with Machines using `fly machine` commands, you can point to an prebuilt image, or build an image from your hand-crafted Dockerfile and source code.

When you `fly launch` an app, you can still point to image or point to Dockerfile, or a Fly Launch scanner may generate an image for you based on your source code.

```

### Build from a Dockerfile
To build the image from a Dockerfile named `Dockerfile`, indicate the current working directory using the `<image>` argument.

```cmd
fly machine run .
```

Use the `--dockerfile` option to specify a Dockerfile with a different name. 

```cmd
fly machine run . --dockerfile Dockerfile-dev
```

### Use an existing image

> 

Reviewer: Add "For example:"? here and above?

### Custom CMD

Override CMD by including the command to run at the end of the `fly machine run` invocation. This example simply spins up a Debian Linux Machine with a `sleep` task to keep it awake; you can shell into it or whatever:

```cmd
fly machine run debian sleep inf
```

### Custom ENTRYPOINT

Override ENTRYPOINT using `--entrypoint`. In this example we use the [`--file-local` option](/docs/machines/run#copy-a-local-file-into-the-machine-file-system) to send an entrypoint script to the Machine and the `--entrypoint` option to run the script:

```cmd
fly machine run debian --file-local /entrypoint.sh=./entrypoint.sh \
                       --entrypoint "/entrypoint.sh" \
                       sleep inf
```

Here's a trivial `entrypoint.sh` you can use with the above example:

```bash
#! /bin/bash

echo "Hello from my Fly Machine"

exec "$@"
```

## Choose the region

Tell the Fly Platform which [region](/docs/reference/regions/) to create the Machine in with the `--region` flag; if for some reason it can't start a new Machine in that region, you'll get an error. If the `--region` flag is omitted, the platform chooses the nearest region to you.

> ## Set hardware specs

Reviewer: do we say hardware specs or machine resources here?

Author: I bet people will find it better with hardware specs, but machine resources is more accurate

---
title: Run a new Machine
objective: Run a new Fly Machine with flyctl
layout: docs
nav: machines
order: 5
---

The [`fly machine run` command](/docs/flyctl/machine-run/) is a tool to configure, build, and run a new Machine in a single guided interaction.

Use `fly machine run` to include Machines built from more than one single Docker image in a Fly App, or to run a one-off or temporary Machine.

Fly Launch features like `fly deploy`, `fly status`, and `fly scale` don't apply to Machines created with `fly machine run`, unless you [add metadata to indicate otherwise](/docs/machines/run#add-metadata-to-the-machine).

Here's what `fly machine run` does for you:
* Checks with the platform for the org and app you've specified, if any, and if needed, guides you through naming a new app
* Creates a Fly App, if applicable
* Gets, or builds, a Docker image to make the Machine from
* Creates the Machine with some default config, plus config you pass to it with flags
* Starts the Machine
* Waits for the Machine to start before declaring success (or failure)

To create a Machine, but not start it, use [`fly machine create`](/docs/flyctl/machine-create/).

> To make changes to a Machine once it's created or run, use [`fly machine update`](/docs/flyctl/machine-create/).

Author: ```suggestion
To make changes to a Machine once it's created or run, use [`fly machine update`](/docs/flyctl/machine-update/).
```
